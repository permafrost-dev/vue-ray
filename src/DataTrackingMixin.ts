export const DataTrackingMixin = {
    data() {
        return {
            trackingRays: {},
            trackingStops: {},
        };
    },
};
