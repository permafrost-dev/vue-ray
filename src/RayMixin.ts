import type { VueRay } from '@/VueRay';
import { rayWrapped } from '@/VueRay';
import { determineComponentNameDuringEvent } from '@/lib/helpers';

const displayLifecycleEvent = (eventName: string, options: Record<string, unknown>, rayInstance: any = null) => {
    rayInstance = rayInstance ?? RayMixin.methods.$ray;

    if (options.type) {
        options = options.type as any;
    }

    // don't display 'unknown' components
    if (!(options?.__file ?? false)) {
        return;
    }

    const componentName: string = determineComponentNameDuringEvent(options);

    let r: VueRay | CallableFunction = rayInstance;

    if (typeof rayInstance === 'function') {
        r = rayInstance();
    }

    (r as VueRay).table([
        `component: <code class="text-green-600 font-bold bold">${componentName}</code>`,
        `lifecycle event: <code class="text-blue-600">${eventName}</code>`,
        `filename: <code class="text-gray-700">${options?.__file ?? 'unknown.js'}</code>`,
    ]);
};

export const RayMixin = {
    beforeMount(component: any, instance: any = null) {
        displayLifecycleEvent('before-mount', component, instance);
    },

    beforeUnmount(component: any, instance: any = null) {
        displayLifecycleEvent('before-unmount', component, instance);
    },

    created(component: any, instance: any = null) {
        displayLifecycleEvent('created', component, instance);
    },

    mounted(component: any, instance: any = null) {
        displayLifecycleEvent('mounted', component, instance);
    },

    unmounted(component: any, instance: any = null) {
        displayLifecycleEvent('unmounted', component, instance);
    },

    updated(component: any, instance: any = null) {
        displayLifecycleEvent('updated', component, instance);
    },

    methods: {
        // eslint-disable-next-line
        $ray(component): (...args: any[]) => VueRay {
            return rayWrapped(component);
        },
    },
};
