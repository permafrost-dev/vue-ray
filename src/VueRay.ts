import { determineComponentNameDuringEvent } from '@/lib/helpers';
import { RayMixin } from '@/RayMixin';
import { format as prettyPrint } from '@permafrost-dev/pretty-format';
import { Ray } from 'node-ray/dist/web';
import { getCurrentInstance, type Ref, watch } from 'vue';

export class VueRay extends Ray {
    public component: any = { $data: {}, $refs: {} };
    public $watch: CallableFunction = () => {};

    public data(): void {
        if (this.component) {
            let data = Object.assign({}, this.component.data);

            // remove Vue Observers to avoid recursive issues
            delete data['trackingRays'];
            delete data['trackingStops'];

            data = Object.freeze(data);

            this.table([data], 'Data');
        }
    }

    public props(): void {
        if (this.component) {
            this.table([this.component.props]);
        }
    }

    public element(refName: string): VueRay {
        if (this.component?.refs[refName]) {
            return this.html((this.component?.refs[refName] as any).outerHTML);
        }

        return this;
    }

    public track(name: string): void {
        if (this.component && typeof this.component.trackingRays[name] === 'undefined') {
            this.component.trackingRays[name] = RayMixin.methods.$ray(this.component)(this.component.data['name']);

            const onTrackedUpdate = (value: any, previousValue: any) => {
                this.component.trackingRays[name] = this.component.trackingRays[name].sendCustom(
                    [
                        `<code style="font-weight: bold;">tracking: </code><code style="color: #16a34a;">${determineComponentNameDuringEvent(this.component.type)}.data.${name}</code>`,
                        `<hr width="33%" align="left" /><code style="color: #94a3b8;">new value: </code><code class="text-gray-500 bold font-bold">${prettyPrint(value)}</div>`,
                        `                               <code style="color: #94a3b8;">old value: </code><code class="text-gray-700">${prettyPrint(previousValue)}</code>`,
                        `<hr width="33%" align="left" /><code style="color: #94a3b8;">changed at: </code><code style="color: #60a5fa;">${new Date().toISOString()}</div>`,
                    ].join('<br>'),
                    'Tracking Data',
                );
            };

            this.component.trackingStops[name] = this.$watch(() => this.component.data[name], onTrackedUpdate, { deep: true });
        }
    }

    public untrack(name: string): void {
        if (this.component && typeof this.component.trackingStops[name] !== 'undefined') {
            this.component.trackingStops[name]();
            delete this.component.trackingStops[name];
        }

        if (this.component && typeof this.component.trackingRays[name] !== 'undefined') {
            delete this.component.trackingRays[name];
        }
    }

    public unwatch(name: string): void {
        if (this.component && typeof this.component.trackingStops[name] !== 'undefined') {
            this.component.trackingStops[name]();
            delete this.component.trackingStops[name];
        }

        if (this.component && typeof this.component.trackingRays[name] !== 'undefined') {
            delete this.component.trackingRays[name];
        }
    }

    public watch(name: string, ref: Ref<any>): void {
        if (this.component && typeof this.component.trackingRays[name] === 'undefined') {
            this.component.trackingRays[name] = RayMixin.methods.$ray(this.component)(ref.value);

            const onTrackedUpdate = (value: any, previousValue: any) => {
                this.component.trackingRays[name] = this.component.trackingRays[name].sendCustom(
                    [
                        `<code style="font-weight: bold;">watching: </code><code style="color: #16a34a;">${determineComponentNameDuringEvent(this.component.type)} âžž Ref&lt;${name}&gt;</code>`,
                        `<hr width="33%" align="left" /><code style="color: #94a3b8;">new value: </code><code class="text-gray-500 bold font-bold">${prettyPrint(value)}</div>`,
                        `                               <code style="color: #94a3b8;">old value: </code><code class="text-gray-700">${prettyPrint(previousValue)}</code>`,
                        `<hr width="33%" align="left" /><code style="color: #94a3b8;">changed at: </code><code style="color: #60a5fa" class="text-gray-500 bold font-bold">${new Date().toISOString()}</div>`,
                    ].join('<br>'),
                    'Watching Ref',
                );
            };

            this.component.trackingStops[name] = this.$watch(ref, onTrackedUpdate, { deep: true });
        }
    }

    public static useDefaultSettings(settings: any) {
        return super.useDefaultSettings(settings);
    }

    public send(...args: any[]) {
        return super.send(...args);
    }
}

export const ray = (...args: any[]) => {
    const result = VueRay.create() as VueRay;
    result.component = getCurrentInstance();
    result.$watch = watch;

    if (!args.length) {
        return result;
    }

    result.send(...args);

    return result;
};

/**
 * Returns a ray() function that is wrapped with the component instance
 *
 * @param component
 * @returns {(...args: any[]) => VueRay}
 */
// eslint-disable-next-line no-unused-vars
export function rayWrapped(component: any): (...args: any[]) => VueRay {
    // biome-ignore lint/complexity/useArrowFunction: <explanation>
    return function (...args: any[]) {
        const result = VueRay.create() as VueRay;

        result.component = component;
        result.$watch = watch;

        if (!args.length) {
            return result;
        }

        result.send(...args);

        return result;
    };
}
