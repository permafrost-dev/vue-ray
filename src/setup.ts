import { RayMixin } from '@/RayMixin';
import { getCurrentInstance, ref } from 'vue';

export function raySetup() {
    const component = getCurrentInstance();

    for (const key of Object.keys(RayMixin.methods)) {
        component.data[key] = RayMixin.methods[key];
    }

    const rayInstance = ref(component.data.$ray);

    return rayInstance;
}
