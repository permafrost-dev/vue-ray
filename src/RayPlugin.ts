import { initializeOptions } from '@/InitializeOptions';
import { createPackageMetaProperty } from '@/lib/helpers';
import { rayWrapped } from '@/VueRay';
import type { App } from 'vue';

export const RayPlugin = {
    install: (app: App, options: any) => {
        createPackageMetaProperty(app.config.globalProperties);
        initializeOptions(options, app.config);

        app.provide('ray', options);
        app['$ray'] = rayWrapped(app);
    },
};
