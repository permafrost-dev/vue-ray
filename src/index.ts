import { VueRay, ray } from '@/VueRay';
import { RayMixin } from '@/RayMixin';
import { DataTrackingMixin } from '@/DataTrackingMixin';
import { initializeOptions } from '@/InitializeOptions';
import { raySetup } from '@/setup';

export { VueRay, ray, RayMixin, DataTrackingMixin, initializeOptions, raySetup };
