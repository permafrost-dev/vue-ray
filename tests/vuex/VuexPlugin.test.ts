/* eslint-disable no-unused-vars */
/* eslint-disable no-undef */

import { VueRay } from '../../src/VueRay';
import { FakeClient } from '../TestClasses/FakeClient';
import { FakeVuexStore } from '../TestClasses/FakeVuexStore';
import { VuexPlugin } from '../../src/vuex/VuexPlugin';

let ray: VueRay, client: FakeClient, store: FakeVuexStore;

beforeEach(() => {
    client = new FakeClient(3000, 'localhost');
    // @ts-ignore
    ray = VueRay.create(client, 'fakeUuid');

    ray.component = {
        $data: { one: 1 },
        $props: { title: 'abc' },
        $refs: { test1: { innerHTML: '<em>123</em>' } },
    };

    store = new FakeVuexStore();
});

describe('VuexPlugin:', () => {
    it('tracks the vuex state', () => {
        const plugin = VuexPlugin({ trackState: true }, () => ray);
        plugin(store);

        store.callMutationSubscriptions('testMutation', { a: 1 }, { a: 123, b: 456 });

        expect(client.sentRequests).toMatchSnapshot();
    });

    it('logs mutation events', () => {
        const plugin = VuexPlugin({ logMutations: true, loggedMutationColor: 'blue' }, () => ray);
        plugin(store);

        store.callMutationSubscriptions('testMutation', { a: 1 }, { a: 123, b: 456 });

        expect(client.sentRequests).toMatchSnapshot();
    });

    it('logs action events', () => {
        const plugin = VuexPlugin({ logActions: true, loggedActionColor: 'blue' }, () => ray);
        plugin(store);

        store.callActionSubscriptions('testAction', 12, { a: 123, b: 456 });

        expect(client.sentRequests).toMatchSnapshot();
    });
});
