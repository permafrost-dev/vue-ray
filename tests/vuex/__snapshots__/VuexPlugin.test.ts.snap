// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VuexPlugin: logs action events 1`] = `
[
  {
    "meta": [],
    "payloads": [
      {
        "content": {
          "label": "Vuex",
          "values": {
            "Payload": 12,
            "Vuex Event": "<span class="pr-1 text-blue-600">action</span> testAction",
          },
        },
        "origin": {
          "file": "web.cjs.js",
          "function_name": "xxxx",
          "hostname": "remote",
          "line_number": 999,
        },
        "type": "table",
      },
    ],
    "uuid": "fakeUuid",
  },
  {
    "meta": [],
    "payloads": [
      {
        "content": {
          "color": "blue",
        },
        "origin": {
          "file": "web.cjs.js",
          "function_name": "xxxx",
          "hostname": "remote",
          "line_number": 999,
        },
        "type": "color",
      },
    ],
    "uuid": "fakeUuid",
  },
  {
    "meta": [],
    "payloads": [
      {
        "content": {},
        "origin": {
          "file": "web.cjs.js",
          "function_name": "xxxx",
          "hostname": "remote",
          "line_number": 999,
        },
        "type": "hide",
      },
    ],
    "uuid": "fakeUuid",
  },
]
`;

exports[`VuexPlugin: logs mutation events 1`] = `
[
  {
    "meta": [],
    "payloads": [
      {
        "content": {
          "label": "Vuex",
          "values": {
            "Payload": "<code style="font-size: 0.8rem!important;"><span style="font-size: 0.8rem!important;" class="text-yellow-600 bold">Object</span>&nbsp;<code style="font-size: 0.8rem!important;" class="text-gray-600 "><span style="font-size: 0.8rem!important;" class="text-orange-400 bold">{</span><br>&nbsp;&nbsp;&nbsp;&nbsp;1<code style="font-size: 0.8rem!important;" class="bold text-green-600 p-0">"a"</code><span style="font-size: 0.8rem!important;" class="text-orange-400 bold">:&nbsp;</span>1<span style="font-size: 0.8rem!important;" class="text-orange-400 bold">,</span><br><span style="font-size: 0.8rem!important;" class="text-orange-400 bold">}</span></code></code>",
            "Vuex Event": "<span class="pr-1 text-green-600">mutation</span> testMutation",
          },
        },
        "origin": {
          "file": "web.cjs.js",
          "function_name": "xxxx",
          "hostname": "remote",
          "line_number": 999,
        },
        "type": "table",
      },
    ],
    "uuid": "fakeUuid",
  },
  {
    "meta": [],
    "payloads": [
      {
        "content": {
          "color": "blue",
        },
        "origin": {
          "file": "web.cjs.js",
          "function_name": "xxxx",
          "hostname": "remote",
          "line_number": 999,
        },
        "type": "color",
      },
    ],
    "uuid": "fakeUuid",
  },
  {
    "meta": [],
    "payloads": [
      {
        "content": {},
        "origin": {
          "file": "web.cjs.js",
          "function_name": "xxxx",
          "hostname": "remote",
          "line_number": 999,
        },
        "type": "hide",
      },
    ],
    "uuid": "fakeUuid",
  },
]
`;

exports[`VuexPlugin: tracks the vuex state 1`] = `
[
  {
    "meta": [],
    "payloads": [
      {
        "content": {
          "content": "<code style="font-size: 0.8rem!important;">VuexState&nbsp;<code style="font-size: 0.8rem!important;" class="text-gray-600 "><span style="font-size: 0.8rem!important;" class="text-orange-400 bold">{</span><br>&nbsp;&nbsp;&nbsp;&nbsp;1<code style="font-size: 0.8rem!important;" class="bold text-green-600 p-0">"state"</code><span style="font-size: 0.8rem!important;" class="text-orange-400 bold">:&nbsp;</span><span style="font-size: 0.8rem!important;" class="text-yellow-600 bold">Object</span>&nbsp;<span style="font-size: 0.8rem!important;" class="text-orange-400 bold">{</span><br>&nbsp;&nbsp;&nbsp;&nbsp;1&nbsp;&nbsp;&nbsp;&nbsp;1<code style="font-size: 0.8rem!important;" class="bold text-green-600 p-0">"a"</code><span style="font-size: 0.8rem!important;" class="text-orange-400 bold">:&nbsp;</span>123<span style="font-size: 0.8rem!important;" class="text-orange-400 bold">,</span><br>&nbsp;&nbsp;&nbsp;&nbsp;1&nbsp;&nbsp;&nbsp;&nbsp;1<code style="font-size: 0.8rem!important;" class="bold text-green-600 p-0">"b"</code><span style="font-size: 0.8rem!important;" class="text-orange-400 bold">:&nbsp;</span>456<span style="font-size: 0.8rem!important;" class="text-orange-400 bold">,</span><br>&nbsp;&nbsp;&nbsp;&nbsp;1<span style="font-size: 0.8rem!important;" class="text-orange-400 bold">}</span><span style="font-size: 0.8rem!important;" class="text-orange-400 bold">,</span><br><span style="font-size: 0.8rem!important;" class="text-orange-400 bold">}</span></code></code>",
          "label": "HTML",
        },
        "origin": {
          "file": "web.cjs.js",
          "function_name": "xxxx",
          "hostname": "remote",
          "line_number": 999,
        },
        "type": "custom",
      },
    ],
    "uuid": "fakeUuid",
  },
]
`;
